source/portable/mbedtls/core_pkcs11_mbedtls.c

```diff
@@ -4008,6 +4012,176 @@ CK_DECLARE_FUNCTION( CK_RV, C_Sign )( CK_SESSION_HANDLE hSession,
 }
 /* @[declare_pkcs11_mbedtls_c_sign] */
 
+/**
+ * @brief Helper function for cleaning up a verify operation for an HMAC operation.
+ * @param[in] pxSession   Pointer to a valid PKCS #11 session.
+ */
+static void prvVerifyInitHMACCleanUp( P11Session_t * pxSession )
+{
+    pxSession->xHMACKeyHandle = CK_INVALID_HANDLE;
+    mbedtls_md_free( &pxSession->xHMACSecretContext );
+}
+
+/**
+ * @brief Helper function for initializing a verify operation for SHA256-HMAC.
+ * @param[in] pxSession         Pointer to a valid PKCS #11 session.
+ * @param[in] hKey              HMAC secret key handle.
+ * @param[in] pucKeyData        HMAC secret key data.
+ * @param[in] ulKeyDataLength   HMAC key Size.
+ */
+static CK_RV prvVerifyInitSHA256HMAC( P11Session_t * pxSession,
+                                      CK_OBJECT_HANDLE hKey,
+                                      CK_BYTE_PTR pucKeyData,
+                                      CK_ULONG ulKeyDataLength )
+{
+    CK_RV xResult = CKR_OK;
+    int32_t lMbedTLSResult = 0;
+    const mbedtls_md_info_t * pxMdInfo = NULL;
+
+    /* Grab the verify mutex.  This ensures that no signing operation
+     * is underway on another thread where modification of key would lead to hard fault.*/
+    mbedtls_md_init( &pxSession->xHMACSecretContext );
+    pxMdInfo = mbedtls_md_info_from_type( MBEDTLS_MD_SHA256 );
+
+    if( pxMdInfo == NULL )
+    {
+        LogError( ( "Failed to initialize verify operation. "
+                    "mbedtls_md_info_from_type failed. Consider "
+                    "double checking the mbedtls_md_type_t object "
+                    "that was used." ) );
+        xResult = CKR_FUNCTION_FAILED;
+        prvVerifyInitHMACCleanUp( pxSession );
+    }
+
+    if( xResult == CKR_OK )
```

Comment:
I can't nest it to stay within the complexity function limits. Also I think a flatter if sequence is easier to read

Reply:
If you remove the lMbedTLSResult variable and check the return values directly, you can keep the complexity level at 8.

Something like this is what I had in mind

```
static CK_RV prvVerifyInitSHA256HMAC( P11Session_t * pxSession,
                                      CK_OBJECT_HANDLE hKey,
                                      CK_BYTE_PTR pucKeyData,
                                      CK_ULONG ulKeyDataLength )
{
    CK_RV xResult = CKR_GENERAL_ERROR;
    const mbedtls_md_info_t * pxMdInfo = NULL;

    /* Grab the verify mutex.  This ensures that no signing operation
     * is underway on another thread where modification of key would lead to hard fault.*/
    mbedtls_md_init( &pxSession->xHMACSecretContext );
    pxMdInfo = mbedtls_md_info_from_type( MBEDTLS_MD_SHA256 );

    if( pxMdInfo == NULL )
    {
        LogError( ( "Failed to initialize verify operation. "
                    "mbedtls_md_info_from_type failed. Consider "
                    "double checking the mbedtls_md_type_t object "
                    "that was used." ) );
        xResult = CKR_FUNCTION_FAILED;
        prvVerifyInitHMACCleanUp( pxSession );
    }

    else if( mbedtls_md_setup( &pxSession->xHMACSecretContext,
                                           pxMdInfo,
                                           PKCS11_USING_HMAC ) )
    {
        LogError( ( "Failed to initialize verify operation. "
                    "mbedtls_md_setup failed: mbed TLS error = %s : %s.",
                    mbedtlsHighLevelCodeOrDefault( lMbedTLSResult ),
                    mbedtlsLowLevelCodeOrDefault( lMbedTLSResult ) ) );
        prvVerifyInitHMACCleanUp( pxSession );
        xResult = CKR_KEY_HANDLE_INVALID;
    }

    else if( mbedtls_md_hmac_starts( &pxSession->xHMACSecretContext,
                                                 pucKeyData, ulKeyDataLength ) )
    {
        LogError( ( "Failed to initialize verify operation. "
                    "mbedtls_md_setup failed: mbed TLS error = %s : %s.",
                    mbedtlsHighLevelCodeOrDefault( lMbedTLSResult ),
                    mbedtlsLowLevelCodeOrDefault( lMbedTLSResult ) ) );
        prvVerifyInitHMACCleanUp( pxSession );
        xResult = CKR_KEY_HANDLE_INVALID;
    }
    else
    {
        xResult = CKR_OK;
        pxSession->xHMACKeyHandle = hKey;
    }

    return xResult;
}
```

It's okay to keep it as-is though since the rest of this file is structured in a similar way.

Reply:
I personally prefer this style as well, but it goes against the formatting guidelines (spoke or unspoken) as I have been asked to keep results in a variable, and then check the if statement